nixos/modules/installer/cd-dvd/iso-image.nix

```diff
@@ -344,12 +344,30 @@ let
       ${pkgs.dosfstools}/sbin/fsck.vfat -vn "$out"
     ''; # */
 
-  targetArch = if pkgs.stdenv.isi686 then
-    "ia32"
-  else if pkgs.stdenv.isx86_64 then
-    "x64"
-  else
-    throw "Unsupported architecture";
+  # Name used by UEFI for architectures.
+  targetArch =
+    if pkgs.stdenv.isi686 then
+      "ia32"
+    else if pkgs.stdenv.isx86_64 then
+      "x64"
+    else if pkgs.stdenv.isAarch64 then
+      "aa64"
+    else
+      throw "Unsupported architecture";
+
+  # Name used by GRUB for architectures.
+  grubTargetArch =
+    if pkgs.stdenv.isi686 then
+      "i386"
+    else if pkgs.stdenv.isx86_64 then
+      "x86_64"
+    else if pkgs.stdenv.isAarch64 then
+      "arm64"
+    else
+      throw "Unsupported architecture";
+
+  # Syslinux (and isolinux) only supports x86-based architectures.
+  isx86Compatible = pkgs.stdenv.isi686 || pkgs.stdenv.isx86_64;
```

Comment:
Nit: Maybe this could be something like `biosBootable` or `x86BiosBootable`? It should be possible to build EFI-only x86 images after all.

Reply:
Hmm, while you're right for the documented purpose, it's also not right for one feature, whether or not to add the `efi_uga` module for GRUB. I may as well change how it is added by instead checking for the existence of said module; it isn't entirely right either to say it's an x86 thing.

Reply:
Does attempting to load `efi_uga` cause any problems in practice? IIRC it doesn't since `install-grub.pl` will put it into the final `grub.cfg` anyway on `nixos-rebuild switch` and it has worked with the community box.

Reply:
The issue here is that efi_uga isn't built. Removing the newly added `if`.

```
Adding additional modules:
 - efi_uga
/nix/store/d07hfl9whmwjhj4pnajk1yz0fnr51j91-grub-2.02/bin/grub-mkimage: error: cannot open `/nix/store/d07hfl9whmwjhj4pnajk1yz0fnr51j91-grub-2.02/lib/grub/arm64-efi/efi_uga.mod': No such file or directory.
builder for '/nix/store/vhwh044bcn4fp1f7hs5hpiixw6m0vhdy-efi-directory.drv' failed with exit code 1
cannot build derivation '/nix/store/17vd0jkmdh7aq3fbdvkcarqc5nm63whh-nixos-19.03.git.ad27b06-aarch64-linux.iso.drv': 1
dependencies couldn't be built
error: build of '/nix/store/17vd0jkmdh7aq3fbdvkcarqc5nm63whh-nixos-19.03.git.ad27b06-aarch64-linux.iso.drv' failed
```

So it's not an issue of attempting to load `efi_uga`, but more of an issue where since it doesn't exist, `grub-mkimage` gets mad. I am not putting the other modules in similar conditionals to ensure they exist. If they don't then something is up with the grub build; either grub was updated and a module was removed (which may need manual intervention) or the build was broken.